engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java

```diff
@@ -0,0 +1,108 @@
+package org.enso.interpreter.runtime.builtin;
+
+import com.oracle.truffle.api.dsl.Bind;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Idempotent;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.node.expression.builtin.Builtin;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * Base class for every Enso builtin object. Not type. Note that base class for a builtin type is
+ * {@link Builtin}.
+ *
+ * <p>In other words, this class represents an object of builtin type in a similar way that {@link
+ * org.enso.interpreter.runtime.data.atom.Atom} represents an object of a non-builtin type.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+public abstract class BuiltinObject extends EnsoObject {
+
+  @ExportMessage
+  public final boolean hasType() {
+    return true;
+  }
+
+  /**
+   * Returns the name of the builtin as saved inside {@link Builtins#builtinsByName}. Not fully
+   * qualified.
+   *
+   * @return
+   */
+  protected abstract String builtinName();
+
+  protected final Type getBuiltinType() {
+    return GetType.uncached(this);
+  }
+
+  /**
+   * Must return false, otherwise if a builtin object is passed to a host method that has a single
+   * {@code Object} argument, host interop would convert the builtin object to a {@code Map} with
+   * all its members. Even if the builtin object is, e.g., a number of a date.
+   *
+   * <p>Must return false as long as all our stdlib Java methods accept {@code Object} and not
+   * {@link org.graalvm.polyglot.Value} as arguments comming from Enso.
+   */
+  @ExportMessage
+  public final boolean hasMembers() {
+    return false;
+  }
+
+  @ExportMessage
+  public final Object getMembers(boolean includeInternal) throws UnsupportedMessageException {
+    throw UnsupportedMessageException.create();
+  }
+
+  @ExportMessage
+  public final boolean hasMetaObject() {
+    return true;
+  }
+
+  @ExportMessage(name = "getType", library = TypesLibrary.class)
+  @ExportMessage(name = "getMetaObject", library = InteropLibrary.class)
+  public static class GetType {
```

Comment:
I guess this class cannot be _package private_ as subclasses need to see the messages. Can the class be `final`? Probably not. Maybe it can be `protected` class? Can we have _package private_ constructor?

Reply:
It cannot be protected. It can be `final` and have package-private constructor - 7a2ea7c0a5ecb45c3b6130ca82b39bd9054f8b97